pkg/aws/eks.go

```diff
@@ -0,0 +1,76 @@
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/service/eks"
+	"github.com/stretchr/testify/require"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"sigs.k8s.io/aws-iam-authenticator/pkg/token"
+)
+
+type EKSClient interface {
+	DescribeCluster(context.Context, *eks.DescribeClusterInput, ...*eks.Options) (*eks.DescribeClusterOutput, error)
+}
+
+type GetEKSTokenInput struct {
+	ClusterName string
+}
+
+type GetEKSTokenOutput struct {
+	Token    string
+	Endpoint string
+	CAData   []byte
+}
+
+func GetEKSClusterE(t *testing.T, ctx context.Context, client EKSClient, input *GetEKSTokenInput) (output *GetEKSTokenOutput, err error) {
+	describeClusterInput := &eks.DescribeClusterInput{
+		Name: &input.ClusterName,
+	}
+	describeOutput, err := client.DescribeCluster(ctx, describeClusterInput)
+	require.Nil(t, err)
+	require.NotNil(t, describeOutput)
+	output = &GetEKSTokenOutput{}
+	output.Endpoint = *describeOutput.Cluster.Endpoint
+	output.CAData, err = base64.StdEncoding.DecodeString(*describeOutput.Cluster.CertificateAuthority.Data)
+	return
+}
+
+type GetEKSClientsetInput struct {
+	ClusterName     string
+	ClusterEndpoint string
+	GetWithOptions  func(*token.GetTokenOptions) (*token.Token, error)
+	NewConfigFunc   func(*rest.Config) (*kubernetes.Clientset, error)
+	ClusterCAData   []byte
+}
+
+type Generator interface {
+	GetWithOptions(*token.GetTokenOptions) (*token.Token, error)
+}
+
+type Kubernetes interface {
+	NewForConfig(*rest.Config) (*kubernetes.Clientset, error)
+}
+
+func GetEKSClientsetE(t *testing.T, ctx context.Context, input *GetEKSClientsetInput) (clientset *kubernetes.Clientset, err error) {
```

Comment:
I'm a little conflicted on whether this belongs here or in the `k8s` package. Also, I don't really love how I'm having the input struct include funcs; I couldn't figure out a better way to allow easy mocking of the methods that generate the token and create the `clientset` object.

Reply:
Is your issue with the fact that you are passing functions, or the way you are passing functions `GetEKSClientsetE ` (the API)? 

If the issue is passing functions, I wouldn't be too concerned. Passing functions, particularly to setup or configuration functions, can be very common in Golang, particularly if using functional options (see below).

If the issue is this function's API, I've found functional options can often help in situations like this: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis.

Reply:
>Passing functions, particularly to setup or configuration functions, can be very common in Golang, particularly if using functional options (see below).

Ah good, that makes me feel a little better about this. It strikes me as perhaps a bit overly complex from the perspective of the consumer, given that a goal of Infratest is to make writing Go tests for infrastructure easier, and by exposing this API I feel like we're failing there a bit. But perhaps if this pattern is more common than my somewhat limited Go experience tells me, that's not a valid concern.

Thanks for the link on functional options! I'll take a look there and see if that's a better / more idiomatic pattern to follow.